### PR TITLE
Add UPP

### LINKED ISSUE #6545: new module: upp/align
### Is there an existing module for this?

- [X] I have searched for the existing module

### Is there an open PR for this?

- [X] I have searched for existing PRs

### Is there an open issue for this?

- [X] I have searched for existing issues

### Further Information

_No response_

### Are you going to work on this?

- [X] If I'm planning to work on this module, I added myself to the `Assignees` to facilitate tracking who is working on the module

### PR DESCRIPTION
**Pull request by luisas**
## PR checklist
Add module UPP/align
Closes #6545

- [x] This comment contains a description of changes (with reason).
- [x] If you've fixed a bug or added code that should be tested, add tests!
- [x] If you've added a new tool - have you followed the module conventions in the [contribution docs](https://github.com/nf-core/modules/tree/master/.github/CONTRIBUTING.md)
- [x] If necessary, include test data in your PR.
- [x] Remove all TODO statements.
- [x] Emit the `versions.yml` file.
- [x] Follow the naming conventions.
- [x] Follow the parameters requirements.
- [x] Follow the input/output options guidelines.
- [x] Add a resource `label`
- [x] Use BioConda and BioContainers if possible to fulfil software requirements.
- Ensure that the test works with either Docker / Singularity. Conda CI tests can be quite flaky:
  - For modules:
    - [x] `nf-core modules test <MODULE> --profile docker`
    - [ ] `nf-core modules test <MODULE> --profile singularity`
    - [ ] `nf-core modules test <MODULE> --profile conda`
  - For subworkflows:
    - [ ] `nf-core subworkflows test <SUBWORKFLOW> --profile docker`
    - [ ] `nf-core subworkflows test <SUBWORKFLOW> --profile singularity`
    - [ ] `nf-core subworkflows test <SUBWORKFLOW> --profile conda`

### CODE REVIEW

File: modules/nf-core/upp/align/main.nf [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+    "${task.process}":
+        upp: \$(run_upp.py -v | grep "run_upp" | cut -f2 -d" ")
+        pigz: \$(echo \$(pigz --version 2>&1) | sed 's/^.*pigz\\w*//' ))
+    END_VERSIONS
+    """
+
+    stub:
+    def prefix = task.ext.prefix ?: "${meta.id}"
+    """
+    touch ${prefix}.aln${compress ? '.gz':''}
```
Comment (SPPearce) [inline modules/nf-core/upp/align/main.nf:52]:
```suggestion
    echo | gzip > ${prefix}.aln${compress ? '.gz':''}
```
Nf-test doesn't like .gz files that aren't actually zipped

File: modules/nf-core/upp/align/meta.yml [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+      description: |
+        Groovy Map containing tree information
+        e.g. `[ id:'test_tree']`
+  - tree:
+      type: file
+      description: Input guide tree in Newick format
+      pattern: "*.{dnd}"
+  - compress:
+      type: boolean
+      description: Flag representing whether the output MSA should be compressed. Set to true to enable/false to disable compression. Compression is handled by passing '-gz' to FAMSA along with any other options specified in task.ext.args.
```
Comment (SPPearce) [inline modules/nf-core/upp/align/meta.yml:38]: Is this correct for this module?

File: modules/nf-core/upp/align/tests/main.nf.test [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+                    input[2] = false
+                    """
+            }
+        }
+
+        then {
+            assertAll(
+                    { assert process.success },
+                    { assert snapshot(process.out.alignment).match("alignment - uncompressed")},
+                    { assert snapshot(process.out.versions).match("versions_uncomp") }
```
Comment (SPPearce) [inline modules/nf-core/upp/align/tests/main.nf.test:32]:
```suggestion
                    { assert snapshot(process.out).match() }
```

File: modules/nf-core/upp/align/tests/main.nf.test [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+                    input[1] = [[:],[]]
+                    input[2] = false
+                    """
+            }
+        }
+
+        then {
+            assertAll(
+                    { assert process.success },
+                    { assert snapshot(process.out.alignment).match("alignment - uncompressed")},
```
Comment (SPPearce) [inline modules/nf-core/upp/align/tests/main.nf.test:31]: Sorry, on my phone and can't figure out how to suggest over two lines at once
```suggestion
```

**Review by SPPearce — COMMENTED** [review]
Looks generally fine. 
Can you please just snapshot the whole of process.out for all three tests, just neater that way when everything is stable

File: modules/nf-core/upp/align/main.nf [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+        $args \\
+        -x $task.cpus \\
+        -s ${fasta} \\
+        -d . \\
+        -o ${prefix} \\
+        -p ./upp-temporary
+
+    mv ${prefix}_alignment.fasta ${prefix}.aln
+
+    # compress both output files
```
[the remark this answers is not in the document]
Comment (luisas) [reply modules/nf-core/upp/align/main.nf:37]: uh right!

File: modules/nf-core/upp/align/meta.yml (hunk 2 above)
Comment (luisas) [reply modules/nf-core/upp/align/meta.yml:38]: Uh no! thanks!

File: modules/nf-core/upp/align/main.nf (hunk 1 above)
Comment (luisas) [reply modules/nf-core/upp/align/main.nf:52]: ok!

File: modules/nf-core/upp/align/main.nf [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+    """
+
+    stub:
+    def prefix = task.ext.prefix ?: "${meta.id}"
+    """
+
+    if [ "$workflow.containerEngine" = 'singularity' ]; then
+        export CONDA_PREFIX="/opt/conda/"
+        export PASTA_TOOLS_DEVDIR="/opt/conda/bin/"
+    fi
```
Comment (SPPearce) [inline modules/nf-core/upp/align/main.nf:61]:
```suggestion
```

**Review by SPPearce — APPROVED** [review]
Looks good to me if the tests pass.

**Comment by luisas** [comment]
@SPPearce thanks a lot for the review :)

**Pull request merged**